Fixed the Equals bug of floating point number and user-defined struct.

mazong1123 opened the pull request:
This PR disables the fast equal check and fast hash code helper for floating point numbers so that 0.0 and -0.0 can be considered as the same. Same for all NaNs.

This definately can be optimized. For example, we should only opt out 0 and NaN for double/single and let other values use fast equal check and fast hash code helper.

**UPDATE**

The final implementation is as following:

Other than `ContainsPointers` and `IsNotTightlyPacked`, added two new conditions for checking whether a valuetype can go to fast path or not. Following are the details of these 2 conditions:

- Check whether a valuetype contains a Single/Double field. If it does, we cannot go to fast path. Floating point numbers have special `Equals` and `GetHashCode` implementation. We cannot simply compare floating point numbers via bits (e.g. +0.0 should equal to -0.0, but their underlying bits are different).

- Check whether an user-defined valuetype overrides `Equals` or `GetHashCode`. If it does, we cannot go to fast path. Because `Equals` or `GetHashCode` result may be different to bit comparison.

To find Single/Double fields and overridden `Equals`/`GetHashCode`, we start a DFS to go through the whole hierachy of the source valuetype, and cache the result to `MethodTable`.

Fix #11452

mazong1123 commented:
@tannergooding @tarekgh @jkotas 
Current changes fixed the bug. However, the performance may be degraded as I disabled fast check for all floating point numbers other than just for `0` and `NaN`. I create this PR to check with CI and hope someone familiar with the type loader can review my solution - is it on the right way?

If the direction is correct, I may add a `HasFloatingPointNumberField()` method in `EEClass`, and call `HasFloatingPointNumberField()`, check `0` / `NaN` along with `!mt->IsNotTightlyPacked()` to determine whether we can use the fast check. The hard part is I do not know how to add a new flag for `HasFloatingPointNumberField` - seems like all available flags have been taken.

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

tarekgh commented:
CC @AlexGhiondea @joperezr 

Would be nice to wait for @jkotas to review this change.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -1809,7 +1811,7 @@ MethodTableBuilder::BuildMethodTableThrowing(
     {
-        if (bmtFP->NumInstanceFieldBytes != totalDeclaredFieldSize || HasOverLayedField())
+        if (bmtFP->NumInstanceFieldBytes != totalDeclaredFieldSize || HasOverLayedField() || hasFloatingPointNumberField)
             GetHalfBakedClass()->SetIsNotTightlyPacked();
```
This `IsNotTighlyPacked` bit is used to disable JIT "struct promotion" optimizations. I do not think we want to be disabling the "struct promotion" optimization when the struct has float field. You need to find a new bit for this.

mazong1123 replied:
@jkotas I tried to define a new bit for `HasFloatingPointNumberField()` but seems like we already exhausted the available bits. You can see all VMFLAG_*:
https://github.com/dotnet/coreclr/blob/44f57065649af5f8bcbb7c71d827221a7bc1bf7a/src/vm/class.h#L2103

Or should I define another enum struct?

jkotas commented:
The same fundamental bug exists for all fields with a custom Equals operator that do not contain GC reference. float/double are just one instance of this problem. If we are doing something about this, it maybe nice to fix it for all cases, not just float/double.

mazong1123 commented:
@jkotas I may miss some background knowledge about this bug. Could you please give me an example for demonstrating the bug in non-double/float fields?

jkotas commented:
> Could you please give me an example to demonstrate the bug in non-double/float fields?

Actual output: true (NeverEquals.Equals is not called)
Expected output: false

```
using System;

struct NeverEquals
{
    byte _a;

    public override bool Equals(object obj)
    {
        return false;
    }

    public override int GetHashCode()
    {
        return 0;
    }
}

struct MyStruct
{
    NeverEquals _x;
    NeverEquals _y;
}

class Program
{
    static void Main(string[] args)
    {
        MyStruct i = new MyStruct();
        MyStruct j = new MyStruct();

        Console.WriteLine(i.Equals(j));
    }
}
```

jkotas commented:
More realistic example. 

Actual output: false (NeverEquals.Equals is not called) 
Expected output: true

```
using System;

struct BoolAndHint
{
    private byte _a;

    public bool Value => (_a & 1) != 0;
 
    public byte Hint => (byte)(_a >> 1);

    public BoolAndHint(bool value, int hint)
    {
        _a = (byte)((value ? 1 : 0) + (hint << 1));
    }

    public override bool Equals(object obj)
    {
        if (!(obj is BoolAndHint))
            return false;

        BoolAndHint other = (BoolAndHint)obj;
        return Value == other.Value;
    }

    public override int GetHashCode()
    {
        return 0;
    }
}

struct MyStruct
{
    public MyStruct(bool value, int hint)
    {
        _x = new BoolAndHint(value, hint);
    }

    BoolAndHint _x;
}

class Program
{
    static void Main(string[] args)
    {
        MyStruct i = new MyStruct(false, 42);
        MyStruct j = new MyStruct(false, 43);

        Console.WriteLine(i.Equals(j));
    }
}
```

mazong1123 commented:
@jkotas Thanks for the example! Another question is seems like we exhausted the bits for VMFLAG_*:
https://github.com/dotnet/coreclr/blob/44f57065649af5f8bcbb7c71d827221a7bc1bf7a/src/vm/class.h#L2103
I mean we already reach 0x80000000 , I cannot find a larger value that has only 1 bit set...

Should I create a new enum to store my new flag bit?

jkotas commented:
Find an existing unused bit, e.g. `VMFLAG_REMOTING_PROXY_ATTRIBUTE` or `VMFLAG_NOSUPPRESSUNMGDCODEACCESS` are not actually used in .NET Core. It is fine to submit separate cleanup PR to get these bits freed up.

mazong1123 commented:
To be clear, I need two bits. One is for floating/double +0.0, -0.0, NaN scenario, one is for user-defined struct with overrided `Equals`.

`VMFLAG_REMOTING_PROXY_ATTRIBUTE` is used by `HasRemotingProxyAttribute` at
https://github.com/dotnet/coreclr/blob/44f57065649af5f8bcbb7c71d827221a7bc1bf7a/src/vm/class.h#L1414

`HasRemotingProxyAttribute` is used by `RuntimeTypeHandle::HasProxyAttribute` at https://github.com/dotnet/coreclr/blob/c61525b5883e883621f98d44f479b15d790b0533/src/vm/runtimehandles.cpp#L1044

`RuntimeTypeHandle::HasProxyAttribute` is a FCall which is **NOT** used in CoreFX now. So `VMFLAG_REMOTING_PROXY_ATTRIBUTE` is not used by .net core now. I can take this bit.

However, `VMFLAG_NOSUPPRESSUNMGDCODEACCESS` is ued by `HasSuppressUnmanagedCodeAccessAttr`, which has been used in some important function: https://github.com/dotnet/coreclr/blob/9b7e73676b5012406f4cbae8b08628647d9b433d/src/vm/ecall.cpp#L562
I think I should find another available bit.

jkotas commented:
>https://github.com/dotnet/coreclr/blob/9b7e73676b5012406f4cbae8b08628647d9b433d/src/vm/ecall.cpp#L562

This calls `MethodDesc::HasSuppressUnmanagedCodeAccessAttr()` that just returns true in .NET Core, it does not even check any bits.

mazong1123 commented:
@jkotas how about another usage in NDirect::CreateCLRToNativeILStub: https://github.com/dotnet/coreclr/blob/1c9eb774950c98ae65ef5497d805cff2eb565971/src/vm/dllimport.cpp#L5038

Seems like it has some logic related to security.

jkotas commented:
All this is left over code from full .NET Framework partial trust that can be deleted. (It would be nice to do this cleanup in separate PR.)

mazong1123 commented:
@jkotas Thanks. I'll finish this PR first and then clean up these left over code in another PR :)

mazong1123 commented:
@jkotas to fix the overrided Equals issue, I briefly did:

1. Check if current class has overrided Equals.
2. Check if any field has a type that has overrided Equals.
3. If 1 or 2 is true, call GetHalfBakedClass()->SetHasOverridedEquals();
4. In CanCompareBits, check mt->HasOverridedEquals().

To achieve step 2, I have to leverage `pByValueClassCache` to access MethodTable:
 https://github.com/mazong1123/coreclr/blob/36f3c079de2cba5b58839721de435934eab281dd/src/vm/methodtablebuilder.cpp#L4205-L4208

I tried to use `pByValueClass` which looks like more elegant. However, it would cause a segmentation fault for cross gen. Please let me know do I have any elegant way to access MethodTable.

I'll optimize floating point number check later.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -5344,6 +5377,12 @@ MethodTableBuilder::PlaceVirtualMethods()
+                // mazong1123: a valid overrided method.
+                if (!*hasOverridedEqualsMethod)
+                {
+                    *hasOverridedEqualsMethod = strcmp((*it)->GetMethodSignature().GetName(), "Equals") == 0;
```
This can have false positives.

jkotas commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -5344,6 +5377,12 @@ MethodTableBuilder::PlaceVirtualMethods()
+
+                // mazong1123: a valid overrided method.
+                if (!*hasOverridedEqualsMethod)
+                {
```
Comparing method names and signatures is not free. Adding this kind of logic to type loader will make every type pay for it, but Equals will only ever be called on a few types. It would be better to do this lazily, only once somebody calls Equals on the type.

mazong1123 replied:
If I understand correctly, try to find an overrided `Equals` when somebody calls Equals in a ValueType will need to DFS or BFS all the available fields of the types in worse case. If the hierarchy is deep it's very inefficient IMO, for instance:

```
    struct MyStruct2
    {
        public NeverEquals _x;
        public NeverEquals _y;
    }

    struct MyStruct3
    {
        public MyStruct2 _x;
        public MyStruct2 _y;
    }

    struct NeverEquals
    {
        public double _a;

        public override bool Equals(object obj)
        {
            return false;
        }

        public override int GetHashCode()
        {
            return 0;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            NeverEquals i = new NeverEquals();
            NeverEquals j = new NeverEquals();

            Console.WriteLine(i.Equals(j));
        }
    }
```

We need to drill down to `NeverEquals` to find an overrided `Equals`. That makes the "fast equal" no more fast... 

I think we may need some trade-off. For example, something like "if you override any method in a ValueType, `CanCompareBits` will return false`. Because we can easily record whether the user override any method in the hierarchy, we can simply check it in `CanCompareBits`.

Similar to floating point number check. We can easily record whether the user use any floating point number field in the hierarchy, but it's hard to check if any floating point number filed stores `0.0` or `NaN` at runtime (Need a DFS or BFS).

@jkotas please let me know if I missed something. Thanks.

jkotas commented:
> If the hierarchy is deep it's very inefficient IMO

Right, checking for the Equals overrides is not cheap and that's why the result should be cached. The good place to cache the result of this check would be `MethodTableWriteableData::m_dwFlags`

> it's hard to check if any floating point number field stores 0.0 or NaN 

This is close to impossible to check, not just hard.

jkotas commented:
> I think we may need some trade-off.

Yet another way to fix this issue would be to auto-generate the IL code for the Equals method. It would have good steady-state performance for all cases.

mazong1123 commented:
@jkotas  I've roughly implemented the algorithm to determine whether a value type has overridden Equals in its type hierachy.

I'm going to optimize the performance but since I'm not very familiar about the type system code base, it would be great if you can give me a hint:

1. I'm using `ApproxFieldDescIterator` to iterate fields of a type. It's inefficient since memory copy happens each time an iterator popped from the stack. I'm wondering is there any other way to iterate all fields of a type?

2. During the DFS, I'd like to know if a methodtable has already been checked before. Hashtable should be the proper data structure to achieve this. However, I didn't find any suitable hashtable implementation in CLR. Can I use `std::unordered_set`?

jkotas commented:
> I'm using ApproxFieldDescIterator to iterate fields of a type

That's fine. I do not think that the efficiency of ApproxFieldDescIterator is a problem - you can profile it to tell for sure. A similar algorithm is used in number of other places.

Also, value types are expected to be relatively simple and small. Everything will be slow for big complex structs.

> During the DFS, I'd like to know if a methodtable has already been checked previously

`MethodTableWriteableData::m_dwFlags` is a good place to cache this information.

mazong1123 commented:
@jkotas Thanks. I'll optimize the code and profile the performance tommorrow.

mazong1123 commented:
@jkotas I just realized `CanUseFastGetHashCodeHelper` has a similar issue if user overrides `GetHashCode` method. And the 0.0 and NaN issue exists in `CanUseFastGetHashCodeHelper`.

I think I should fix it in a separate PR. What's your opinion?

jkotas commented:
You may want to do the GetHashCode fix first, so that `a.Equals(b) => a.GetHashCode() == b.GetHashCode()` invariant is maintained at all times.

jkotas commented:
(Having both fixes in one PR would be fine with me as well.)

mazong1123 commented:
Now I found another interesting case:

```csharp
    class Cls
    {
        public byte _a;

        public override bool Equals(object obj)
        {
            return false;
        }

        public override int GetHashCode()
        {
            return 0;
        }
    }

    struct MyStruct6
    {
        public Cls _x;
        public Cls _y;
    }

    static void Main(string[] args)
    {
            MyStruct6 i = new MyStruct6();
            MyStruct6 j = new MyStruct6();

            Console.WriteLine(i.Equals(j));
    }
```

The output is `True`. `CanCompareBits` returns false because it contains pointers. That means in a NORMAL compare process, i is considered equal to j. Although it seems like not related to this issue,  I'm wondering is it an expected behavior or a bug?

jkotas commented:
> The output is `True`

Both `_x` and `_y` are null. null is equal to null. I do not see a problem here.

mazong1123 commented:
@jkotas  Never mind. I forgot `Cls` will not be initialized as ValueType.

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,107 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+            }
+
+            fieldMethodTable->SetHasCheckedOverriddenEquals();
+            if (HasOverriddenEqualsWorker(fieldMethodTable))
```
The `SetHasCheckedOverriddenEquals()` call before calling `HasOverriddenEqualsWorker()` does not seem thread-safe to me.

mazong1123 replied:
I think `SetHasCheckedOverriddenEquals()` is thread-safe:
https://github.com/mazong1123/coreclr/blob/294536634f3c5256ca10331c1352242edd37712f/src/vm/methodtable.h#L1263

Are you referring from other perspective?

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,107 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    MethodTable::IntroducedMethodIterator it(mt);
+    for (; it.IsValid(); it.Next())
+    {
+        if (strcmp(it.GetMethodDesc()->GetName(), "Equals") == 0)
```
The class may have an `Equals` method that does not override `Object.Equals`.  In MSIL it is also possible to override a method using a different method name.

mazong1123 replied:
> The class may have an Equals method that does not override Object.Equals
Umm. I may need to check if the Equals is from overriding (if it can be figured out).

>  In MSIL it is also possible to override a method using a different method name.
I think this is out of scope of this issue. If we need to care about this case, we may need to fix from another way - may be generating different IL code for Equals? @jkotas

mazong1123 replied:
Seems like we can just check is it from overriding via `IsVirtual()` here. Because struct is sealed, user has no chance to introduce a new virtual method. So if the method is virtual, it is from parent. 

Note I'm not considering the scenario that user changes the override method name from IL

jkotas replied:
The way to check for overridden method that handles all cases is like this: https://github.com/dotnet/coreclr/search?utf8=%E2%9C%93&q=HasOverriddenStreamMethod&type=

mazong1123 replied:
@jkotas that code sounds like not comparing the signature of the method. I'll take a look to see if it can be applied to `Equals` and `GetHashCode`

jkotas replied:
> code sounds like not comparing the signature of the method

It is checking whether the virtual slot has been overridden. It is the only 100% accurate way to check whether the method has been overridden.

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,163 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+
+    BOOL hasOverriddenMethod = FALSE;
+
+    // Do we have the overriden method in current Type?
```
Please fix overridden -> overridden here and in two other places.

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,163 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+            }
+
+            setHasCheckedOverridden(fieldMethodTable);
+            if (HasOverriddenMethodWorker(fieldMethodTable, methodName))
```
Consider the case when one thread calls `setHasCheckedOverridden`, then preempted before making the `HasOverriddenMethodWorker` call.  Another thread comes and calls `hasCheckedOverridden` on the same type.  That would incorrectly return `true`, even though we have not checked the type for the `Equals` override yet.

mazong1123 replied:
Thanks for the catch! I'll fix it later.

mazong1123 commented:
@jkotas @AntonLapounov I've updated the code. Now I'm using `GetRestoredSlot` to determine whether a method is overridden. Please take a look. Thanks!

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+        FC_RETURN_BOOL(FALSE);
+    }
+
+    FC_RETURN_BOOL(!HasOverriddenEquals(mt));
```
The code that iterates the fields, etc. has to be executed under `HELPER_METHOD_FRAME`. 

https://github.com/dotnet/coreclr/blob/master/Documentation/botr/mscorlib.md and https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md describes the rules for writing FCalls and manually managed code.

mazong1123 replied:
Thanks. I should read these docs first :)

jkotas commented on src/vm/methodtable.h:
```diff
@@ -390,6 +390,12 @@ struct MethodTableWriteableData
+        enum_flag_HasOverriddenEquals       = 0x00000200,     // Is any field type or sub field type overrode Equals
+        enum_flag_HasCheckedOverriddenEquals   = 0x00000400,  // Whether we have checked the overridden Equals
+
+        enum_flag_HasOverriddenGetHashCode  = 0x00000800,     // Is any field type or sub field type overrode GetHashCode
```
Equals and GetHashCode are likely going to be used together. I think you can check for both at the same time. It will save the bit here, and it will be a bit faster for the typically case.

mazong1123 commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    else
+    {
+        // Now we need to do a depth-first-search on the instance fields.
+        Stack<ApproxFieldDescIterator> fieldIterStack;
```
Any particular reason we did dfs recursively? I think it's slower than do it iteratively.

jkotas replied:
The structs are expected to be small. We do not optimize for large structs - where the iterative implementation would start winning.

mazong1123 replied:
OK I think this code should compliant with what you guys have done if we do not have a performance issue here.

mazong1123 commented on src/vm/class.h:
```diff
@@ -1936,7 +1946,8 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
         VMFLAG_HASNONPUBLICFIELDS              = 0x00040000,
-        // unused                              = 0x00080000,
+
+        VMFLAG_HAS_FLOATING_POINT_NUMBER       = 0x00080000, // Indicates whether floating point number fields exist
```
If we have a double or single field, with this bit,  we can avoid a dfs. Otherwise, we need to do dfs for double or single field as what we did for Equals and GetHashCode.

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    auto hasOverriddenEquals = [](MethodTable* mt) -> BOOL {
+        return mt->HasOverriddenEquals();
+    };
+    auto setHasOverriddenEquals = [](MethodTable* mt) -> void {
```
Using function pointers will add CFG (control flow guard) cost on each call.  If we use a single bit as Jan suggested, could we check for both `Equals` and `GetHashCode` overrides at the same time and avoid function pointers?

AntonLapounov commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -3824,6 +3826,14 @@ VOID    MethodTableBuilder::InitializeFieldDescs(FieldDesc *pFieldDescList,
+                {
+                    GetHalfBakedClass()->SetHasFloatingPointNumberField();
+                    hasFloatingPointNumberField = TRUE;
+                }
```
Instead of this `if` I would do just
``` c++
hasFloatingPointNumberField = TRUE;
```
at this place, and add this code after processing all fields:
``` c++
if (hasFloatingPointNumberField) 
    GetHalfBakedClass()->SetHasFloatingPointNumberField(); 
```

mazong1123 commented:
@jkotas @AntonLapounov I've updated the code:
1. Checked overridden `Equals` and `GetHashCode` altogether.
2. Re-implemented DFS recursively.
3. Removed function pointers.
4. Removed `GetHashCode` bits in `methodtable.h`
5. Added `HELPER_METHOD_FRAME` when  necessary (@jkotas please let me know if I made any mistake here).
6. Changed/Added contracts in functions.

@jkotas I still kept `VMFLAG_HAS_FLOATING_POINT_NUMBER` because it can avoid DFS for double or single field. Please let me know if it should be removed from your perspective. Thanks!

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,157 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+        FC_RETURN_BOOL(FALSE);
+    }
+
+    OBJECTREF objRef = ObjectToOBJECTREF(obj);
```
Could you please measure how much slower is the "fast equals" now? I think this will need to be structured differently to avoid the regression.

jkotas commented:
> @jkotas I still kept VMFLAG_HAS_FLOATING_POINT_NUMBER because it can avoid DFS for double or single field. 

You do not need a special DFS for the double and single fields. You can do it together with the check for overridden GetHashCode/Equals.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,167 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    do
+    {
+        // Try to find the overridden methods among all instance fields recursively.
+        MethodTable* fieldMethodTable = pField->LookupFieldTypeHandle().GetMethodTable();
```
LookupFieldTypeHandle returns the type handle only if the type is loaded already - it can return NULL. You need `GetApproxFieldTypeHandleThrowing`.

mazong1123 commented:
> You do not need a special DFS for the double and single fields. You can do it together with the check for overridden GetHashCode/Equals.

@jkotas Let's say we have a user-defined struct with double fields and overridden `Equals`. If we have `VMFLAG_HAS_FLOATING_POINT_NUMBER` set, we can terminate the process at:https://github.com/mazong1123/coreclr/blob/0dbff3f09f654cc122381ae91e80149680fb9c29/src/vm/comutilnative.cpp#L2762-L2764
Instead of entering `HasOverriddenEqualsOrGetHashCodeMethod`. No DFS happens.

If we do not have this flag, we have to check the double field along with checking `Equals` and `GetHashCode` in DFS, which should be slower.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,167 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    }
+
+    // The type itself did not override Equals or GetHashCode, go for its fields.
+    ApproxFieldDescIterator iter = ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
```
This should be written as:

```
    ApproxFieldDescIterator fieldIterator(pMT, ApproxFieldDescIterator::INSTANCE_FIELDS);
    for (FieldDesc *pFD = fieldIterator.Next(); pFD != NULL; pFD = fieldIterator.Next())
    {
        if (pFD->GetFieldType() == ELEMENT_TYPE_VALUETYPE)
        {
             if (HasOverriddenEqualsOrGetHashCodeMethodWorker(pFD->GetApproxFieldTypeHandleThrowing())
                 ...
        }
        else
             ... check for float/double ...
    }
```

jkotas commented:
> Instead of entering HasOverriddenEqualsOrGetHashCodeMethod. No DFS happens.

Then we will use the slow path and the DFS will happen for every invocation of Equals using reflection: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/ValueType.cs#L50 . Having faster way to get to the slow path is not useful.

mazong1123 commented:
@jkotas I may miss something but in my understanding, if we have double in user-defined struct, we should go to the slower reflection path - because 0.0/NaN cannot be compared correctly in the fast path.

jkotas commented:
> So ultimately, even if we go to the DFS, we should expect CanCompareBits return false if there's any double/float field, which can be returned earlier if we have that flag set.

Right. It is not worth optimizing this case by spreading it into type loader and caching more bits because of it is a slow path anyway.

mazong1123 commented:
@jkotas OK that make sense. I'll include the floating point check in the DFS.

AntonLapounov commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,167 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    FieldDesc* pField = iter.Next();
+    while (pField != NULL
+        && pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+        && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
```
Note that we need to check value type fields only.  The `mt->ContainsPointers()` condition in `CanUseFastGetHashCodeHelper` takes care of class reference fields.

mazong1123 commented:
@jkotas @AntonLapounov unfortunately, the performance degradation rate is around 30% ~ 50% (Except double checking):

**Before**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     52     | 195.401 |   9.302 | 191.486 | 248.350
 perftest.ValueEqualTest.TestDoubleZero                | Duration |    100     |  14.325 |   0.134 |  14.187 |  15.072
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  13.072 |   0.226 |  12.797 |  14.161

**After**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     42     | 241.569 |   6.105 | 234.509 | 266.038
 perftest.ValueEqualTest.TestDoubleZero                | Duration |     63     | 161.349 |   4.954 | 156.572 | 176.262
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  19.094 |   0.464 |  18.457 |  20.307

Note the performance degradation for double is as expected because we do it in the right way (comparing using reflection).

mazong1123 commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test

jkotas commented:
> I found following code slows down the whole program:

Are you running your perf test on release build? The OBJECTREFs are plain pointers in release build. They do not have any overhead.

I think that the CanCompareBits has to written this way to avoid the perf regression:

```
FCIMPL1(FC_BOOL_RET, ValueTypeHelper::CanCompareBits, Object* obj)
{
      _ASSERTE(obj != NULL);
      MethodTable* mt = obj->GetMethodTable();

      if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
          FC_RETURN_BOOL(CanCompareBitsOrUseFastGetHashCode());

      FC_INNER_RETURN(FC_BOOL_RET, CanCompareBitsOrUseFastGetHashCodeHelper(mt));
}
```

Similar for the hashcode.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,166 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+        return mt->CanCompareBitsOrUseFastGetHashCode();
+    }
+
+    if (g_pValueTypeMT == NULL)
```
These caches should not be necessary. The implementation of these methods have caches with very similar perf characteristics already.

The caches were needed for the stream methods because of they need them without entering `HELPER_METHOD_FRAME`. You are calling them inside `HELPER_METHOD_FRAME` here, and so it should not be necessary to cache anything.

mazong1123 replied:
Thanks. I'll change this.

mazong1123 commented:
> Are you running your perf test on release build? The OBJECTREFs are plain pointers in release build. They do not have any overhead.

Yes I'm running on release build. I'll apply your suggestion then test it again.

mazong1123 commented:
Below are the performance measurement:

**Before with warmup**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     53     | 190.266 |   2.394 | 188.475 | 197.607
 perftest.ValueEqualTest.TestDoubleZero                | Duration |    100     |  14.064 |   0.425 |  13.848 |  16.878
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  13.103 |   0.497 |  12.595 |  14.747

**After with warmup**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     42     | 242.081 |   5.684 | 237.043 | 257.798
 perftest.ValueEqualTest.TestDoubleZero                | Duration |     62     | 161.592 |   2.921 | 159.472 | 173.467
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  13.570 |   0.514 |  12.926 |  14.622

**Before without warmup**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     51     | 199.606 |  13.006 | 189.037 | 241.922
 perftest.ValueEqualTest.TestDoubleZero                | Duration |    100     |  14.509 |   0.539 |  13.948 |  15.951
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  13.577 |   1.154 |  12.676 |  17.338


**After without warmup**

 Test Name                                             | Metric   | Iterations | AVERAGE | STDEV.S |     MIN |     MAX
:----------------------------------------------------- |:-------- |:----------:| -------:| -------:| -------:| -------:
 perftest.ValueEqualTest.TestDeepOverriddenEquals      | Duration |     40     | 253.457 |   5.669 | 248.543 | 269.343
 perftest.ValueEqualTest.TestDoubleZero                | Duration |     55     | 182.308 |   4.830 | 178.493 | 194.398
 perftest.ValueEqualTest.TestStructNonOverriddenEquals | Duration |    100     |  13.523 |   0.398 |  13.132 |  14.578

For those structs without double/single or overridden `Equal` (`TestStructNonOverriddenEquals`), the performances are identical.

For those structs with overridden Equals (`TestDeepOverriddenEquals`), the performance after change is slower, which is as expected. Because it goes to the reflection path.

For those structs have double field defined (`TestDoubleZero`), the performance after change is slower, which is as expected. Because it goes to the reflection path.

There's nearly no difference with or without "warmup" for the old code, but after the change, the "warmup" slightly speed up the whole program.

@jkotas The performance regression is probably as expected. Do you think any other scenarios I need to test?

mazong1123 commented:
Now I'm confused about the `GC_TRIGGERS` contract inside FCall. According to following definition:

> #define FCALL_CONTRACT \
    STATIC_CONTRACT_SO_TOLERANT; \
    STATIC_CONTRACT_THROWS; \
    /* FCALLS are a special case contract wise, they are "NOTRIGGER, unless you setup a frame" */ \
    STATIC_CONTRACT_GC_NOTRIGGER; \
    STATIC_CONTRACT_MODE_COOPERATIVE

I thought I need specify `GC_TRIGGERS` as I'm using frames in `GetHashCodeHelper`. However, an assert happened that said 
> GC_TRIGGERS encountered in a GC_NOTRIGGER scope

Then I think I should define "GC_NOTRIGGER" for my helper functions. This time it gives me the same error in `frame.cpp`. Did I misunderstand something? @jkotas

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,180 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    }
+
+    // We've gone through all instance fields. It's time to cache the result.
+    mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
```
This has race. Other thread can see the "checked" bit, without the result bit set. It would be best to set both the checked and result bits in one interlocked operation, ie have method `void SetCanCompareBitsOrUseFastGetHashCode(BOOL value)`

mazong1123 replied:
Thanks. I need to consider the multi-thread scenario. Will fix this.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2797,18 +3002,24 @@ FCIMPL1(INT32, ValueTypeHelper::GetHashCode, Object* objUNSAFE)
-    
-    return hashCode;
+
+    FC_INNER_RETURN(INT32, GetHashCodeHelper(hashCode, pMT, obj));
```
I think you can just use regular `HELPER_METHOD_FRAME_BEGIN` without `FC_INNER_RETURN`. `HELPER_METHOD_FRAME` is used in other places in this method, and so using it in one more place on a slow path won't slow it down.

mazong1123 commented:
@jkotas I found we didn't specify contracts in other helper functions https://github.com/dotnet/coreclr/blob/e0486761d6c019cb696dcd35aeecc031b1d01eef/src/classlibnative/bcltype/objectnative.cpp#L71
So I removed the contracts in CanCompareBitHelper. Now the CI passed.

Also I've implemented a thread-safe `SetCanCompareBitsOrUseFastGetHashCode`:
https://github.com/mazong1123/coreclr/blob/2fa1fb917cd31957766ad3ded1199912a53666e9/src/vm/methodtable.h#L1250

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,173 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+}
+
+static BOOL CanCompareBitsOrUseFastGetHashCodeWorker(MethodTable* mt,
+                                                     MethodTable* valueTypeMT,
```
These do not need to be arguments. You can just fetch them as needed.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,173 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+                                                   MscorlibBinder::GetMethod(METHOD__VALUE_TYPE__GET_HASH_CODE)->GetSlot());;
+}
+
+NOINLINE static BOOL CanCompareBitsHelper(MethodTable* mt, OBJECTREF objRef)
```
This should return `FC_BOOL_RET` to match the outter function.

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,173 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    if (mt->ContainsPointers()
+        || mt->IsNotTightlyPacked())
+    {
+        mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
```
Setting this bit won't help anything here - we are still going to check `ContainsPointers` and `IsNotTightlyPacked` next time through this function. Should the whole check for `ContainsPointers` and `IsNotTightlyPacked` move to CanCompareBitsOrUseFastGetHashCodeWorker?

mazong1123 replied:
Move `ContainsPointers` and `IsNotTightlyPacked` into `CanCompareBitsOrUseFastGetHashCodeWorker` would probably slow down the type that do has pointers or not tightly packed. How about move the `ContainsPointers` and `IsNotTightlyPacked` after following check?:

```cpp
 if (mt->HasCheckedCanCompareBitsOrUseFastGetHashCode())
    {
        FC_RETURN_BOOL(mt->CanCompareBitsOrUseFastGetHashCode());
    }
```

jkotas replied:
> probably slow down the type that do has pointers or not tightly packed checking at the first time

The performance of the first time check for these does not matter because of they will take slow path in steady state.

The most performant way to write these FCalls is to have the code that executes in steady state in the main method, and have all one time initialization in the helper method.

mazong1123 commented:
@jkotas I've updated the code. Now `ContainsPointers` and `IsNotTightlyPacked` are in `CanCompareBitsOrUseFastGetHashCodeWorker`. I updated related code accordingly. Please take a look. Thanks!

> The most performant way to write these FCalls is to have the code that executes in steady state in the main method, and have all one time initialization in the helper method.

This is interesting. I may think about it after finish this PR.

mazong1123 commented:
@dotnet-bot test Tizen armel Cross Release Build

jkotas commented:
@dotnet-bot test Windows_NT x64 corefx_baseline

jkotas commented:
Do you plan to add tests for this as well?

mazong1123 commented:
Yes. But I guess  tests should be added to corefx and enable them after coreclr bits being included. Or should I add tests in coreclr as well?

jkotas commented:
>  tests should be added to corefx

Agree. Thanks!

jkotas commented:
@mazong1123 Could you please squash the commits and write a description that matches the final change?

mazong1123 commented:
@jktoas sure. Will do it later today.

mazong1123 commented:
@jkotas I squashed most of the commits. However, some commits came before I merge commits from master branch to resolve a conflict. If I squash those commits, the merged commits history would be lost. So I kept them. Please let me know if I need to squash those commits as well.

mazong1123 commented:
@dotnet-bot test Windows_NT x64 corefx_baseline

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline

jkotas commented:
> If I squash those commits, the history of the merged commits would be lost

I do not think that the history for this merge is valuable. It can be all squashed into one commit.

mazong1123 commented:
Oops.. I think I need to redo the squash as the merged commits has been lost.

mazong1123 commented:
@jkotas I cherry-picked some commits so I'd like to run tests on my code this morning (I don't have an environment now). Please don't merge it before I finish the tests in case of any regression bug.

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Windows_NT x86 Checked Build and Test (Jit - CoreFx) 
@dotnet-bot test Tizen armel Cross Debug Build

mazong1123 commented:
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Windows_NT x86 corefx_baseline

mazong1123 commented:
@jkotas the code passed my local tests so the logic should be fine. However, the CI started to complain today. The code change would not cause this failure 2 days ago. Do we have any update on CI server? 

```
08:14:46   SGEN: error SGEN1: Missing required command-line argument: The name of the source assembly..
08:14:46   In Development
08:15:04   
08:15:04 
  Assert failure(PID 4212 [0x00001074], Thread: 1256 [0x04e8]): Assertion failed 'src1->OperIs(GT_LONG)' in 'Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationWriter1:Write64_LongEnum(long):ref:this' (IL size 90)
08:15:04   
08:15:04       File: d:\j\workspace\x86_checked_w---92c0aac3\src\jit\lower.cpp Line: 2167
08:15:04       Image: D:\j\workspace\x86_checked_w---92c0aac3\_\fx\bin\testhost\netcoreapp-Windows_NT-Release-x86\dotnet.exe
```

jkotas commented:
There are two problems here:
- New SGen tests are failing in CI. Tracked by https://github.com/dotnet/corefx/issues/23390
- JIT is asserting with `Assertion failed 'src1->OperIs(GT_LONG)'`. @dotnet/jit-contrib Known?

mazong1123 commented:
@jkotas I'll start a PR to test if master branch code has the same issue. If not, it would be introduced by this code change.

mazong1123 commented:
@jkotas Now I'm pretty much sure the JIT assertion failure is not caused by my code change here. See the same failure here: https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_corefx_baseline_prtest/49/
which is from my test PR: https://github.com/dotnet/coreclr/pull/13480

I've updated the test PR to involve most recently changes. If it still fails, I will fire an issue.

But for this PR I think it should be ready to merge.

mazong1123 commented:
@dotnet-bot test OSX10.12 x64 Checked Build and Test

mazong1123 commented:
@jkotas I've filed issue: https://github.com/dotnet/coreclr/issues/13486

mazong1123 commented:
@dotnet-bot test Windows_NT x86 corefx_baseline
@dotnet-bot test this

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,147 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+    BOOL ret = FALSE;
+
+    HELPER_METHOD_FRAME_BEGIN_RET_ATTRIB_1(Frame::FRAME_ATTR_EXACT_DEPTH|Frame::FRAME_ATTR_CAPTURE_DEPTH_2, objRef);
+
```
Nit: extra line

jkotas commented on src/vm/class.h:
```diff
@@ -1398,29 +1398,6 @@ class EEClass // DO NOT CREATE A NEW EEClass USING NEW!
 
 public:
-
-    inline void SetDoesNotHaveSuppressUnmanagedCodeAccessAttr()
```
These diffs are not part of your change - they are in master already. Could you please rebase them away?

jkotas commented on src/vm/comutilnative.cpp:
```diff
@@ -2624,14 +2623,147 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
+        {
+            // Check current field type.
+            MethodTable* fieldMethodTable = pField->GetApproxFieldTypeHandleThrowing().GetMethodTable();
+            if(!CanCompareBitsOrUseFastGetHashCode(fieldMethodTable))
```
Nit: space after it

mazong1123 replied:
I didn't find extra space after this line or at tail. Do you mean I need add space after it?

jkotas replied:
I meant to say add space between `if` and `(`.

mazong1123 replied:
OK got it. I'll update it.

mazong1123 commented:
I saw the CI failure has been fixed https://github.com/dotnet/coreclr/issues/13486#issuecomment-324081359

@jkotas Can this PR  be merged?

jkotas commented:
I have just look at this. Yes, it looks good to me. I have just commented on a few last nits.

mazong1123 commented:
@jkotas I've updated accordingly. Please take a look. Thanks.

jkotas commented:
This change `ValueType.Equals`, but it did not actually fixed `ValueType.GetHashCode`. Opened: https://github.com/dotnet/coreclr/issues/16545
